docs/rule_authors/writing_toolchains.md

```diff
@@ -0,0 +1,130 @@
+passed to it, or where it is fetched from.
+
+The most straightforward thing to do first is to instantiate one of the
+"system" toolchain rules that the prelude offers. Studying the
+toolchains defined by the `system_demo_toolchains` macro in
+`@prelude//toolchains:demo.bzl` is a good way to get started.
+
+For example, here is how one could define a C++ toolchain that builds
+projects in C++23, with warnings as errors and optimizations enabled:
+
+```python
+load("@prelude//toolchains:cxx.bzl", "system_cxx_toolchain")
+
+system_cxx_toolchain(
+    name = "cxx",
+    compiler_type = "clang",
+    cxx_flags = [
+        "-std=c++23",
+        "-Wall",
+        "-Wextra",
+        "-Werror",
+        "-O3",
+    ],
+    visibility = ["PUBLIC"],
+)
+```
+
+One would typically use
+[`select`](https://buck2.build/docs/rule_authors/configurations_by_example/)
+to customize the toolchain e.g. based on the build mode (debug vs
+release) or compiler type.
+
+Note that several toolchains require you to also define a
+`python_bootstrap` toolchain. This is because those toolchains use
+Python scripts (e.g. to generate compilation databases for C++), and the
+prelude gives you control over the Python toolchain used for this
+purpose. The bootstrap Python toolchain can be different from the
+toolchain used by `python_*` rules. `@prelude//toolchains:python.bzl`
+provides a simple `system_python_bootstrap_toolchain` in case you do not
+care about tightly controlling the Python used when building other
+languages.
+
+Completely custom toolchains can also easily be written as rules with
+`is_toolchain_rule = True` which return a `*ToolchainInfo` struct. The
+[prelude's toolchains](https://github.com/facebook/buck2/tree/main/prelude/toolchains),
+including `system_cxx_toolchain` referenced earlier, can serve as
+examples.
+
+## Accessing a toolchain in a build rule implementation
+
+Before going any further, note that using a toolchain is often
+unnecessary. It is perfectly possible for a build rule to reference
+regular targets as dependencies, without the added complexity of
+defining a toolchain for them. For example, one can add an (often
+`default_only`) `attrs.exec_dep` to a build rule's `attrs` to do code
+generation in a custom rule without creating a toolchain for it, as long
+as there is no need to decouple the code generator being used from the
+rule implementation.
+
+With that being said, when a rule must be made generic over a toolchain,
+all that is needed is to add that toolchain as a `toolchain_dep` attr to
+a build rule. Let's consider the following made up rule that would
+simply call a compiler on a source file:
+
+```python
+# We assume that `FooToolchainInfo` is a struct with a `compiler` field.
+def _foo_binary_impl(ctx: AnalysisContext) -> list[Provider]:
+    output = ctx.actions.declare_output(ctx.label.name)
+    ctx.actions.run(
+        [ctx.attrs._compiler, ctx.attrs.source, "-o", output.as_output()],
```

Comment:
Is `ctx.attrs._compiler` correct here? I'd have expected something like `ctx.attrs._foo_toolchain[FooToolchainInfo].compiler`.

Reply:
Oh right, I must have carelessly edited this right before pushing :facepalm: 
I'll fix this when I have a moment, thanks for catching it.